src/include/function/gds/gds_object_manager.h

```diff
@@ -35,6 +35,56 @@ class ObjectBlock {
     std::atomic<uint64_t> nextPosToWrite;
 };
 
+// Pre-allocated array of objects.
+template<typename T>
+class ObjectArray {
+public:
+    ObjectArray(common::offset_t size, storage::MemoryManager* mm, bool initializeToZero = false)
+        : allocation{mm->allocateBuffer(initializeToZero, size * sizeof(T))} {
+        data = std::span<T>(reinterpret_cast<T*>(allocation->getData()), size);
+    }
+
+    void set(common::offset_t pos, T value) {
+        KU_ASSERT(pos < data.size());
+        data[pos] = value;
+    }
+
+    T get(common::offset_t pos) {
+        KU_ASSERT(pos < data.size());
+        return data[pos];
+    }
+private:
+    template<typename U> friend class AtomicObjectArray;
+    std::span<T> data;
+    std::unique_ptr<storage::MemoryBuffer> allocation;
+};
+
+// Pre-allocated array of atomic objects.
+template<typename T>
+class AtomicObjectArray {
+public:
+    AtomicObjectArray(common::offset_t size, storage::MemoryManager* mm, bool initializeToZero = false)
+        : array{ObjectArray<std::atomic<T>>(size, mm, initializeToZero)} {
+    }
+
+    void set(common::offset_t pos, T& value) {
```

Comment:
Let's name these operations setRelaxed so it's clear what level of guarantee the set and get functions give.